File: src/main/scala/leetcode/normal/n705/MyHashSet.scala
package leetcode.normal.n705

import scala.collection.mutable

/**
  * https://leetcode-cn.com/problems/design-hashset/
  * 705. 设计哈希集合
  */
class MyHashSet() {

  /** Initialize your data structure here. */

  val ss = mutable.Set.empty[Int]

  def add(key: Int): Unit = {
    ss.add(key)
  }

  def remove(key: Int): Unit = {
    ss.remove(key)
  }

  /** Returns true if this set contains the specified element */
  def contains(key: Int): Boolean = {
    ss.contains(key)
  }
}

/**
  * Your MyHashSet object will be instantiated and called as such:
  * var obj = new MyHashSet()
  * obj.add(key)
  * obj.remove(key)
  * var param_3 = obj.contains(key)
  */

File: src/main/java/leetcode/normal/j705/MyHashSet.java
package leetcode.normal.j705;

import java.util.LinkedList;

/**
 * Created by CHENXCHEN
 * <a href="https://leetcode.cn/problems/design-hashset/description/">705. 设计哈希集合</a>
 *
 * @author <a href="mailto:chenhuachaoxyz@gmail.com">报时</a>
 */
class MyHashSet {
    int mod = (int) 1e5;
    LinkedList<Integer>[] list;

    public MyHashSet() {
        list = new LinkedList[mod];
    }

    public void add(int key) {
        int sign = key % mod;
        if (list[sign] == null) list[sign] = new LinkedList<>();
        if (list[sign].stream().noneMatch(x -> x == key))
            list[sign].add(key);
    }

    public void remove(int key) {
        int sign = key % mod;
        LinkedList<Integer> linked = list[sign];
        if (linked != null) {
            linked.removeFirstOccurrence(key);
        }
    }

    public boolean contains(int key) {
        int sign = key % mod;
        LinkedList<Integer> linked = list[sign];
        return linked != null && linked.stream().anyMatch(x -> x == key);
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */